### ISSUE TITLE
True != True

### ISSUE THREAD

**Issue by mengesb** [hashicorp/terraform#6447]
Terraform variables behave differently when overridden from `terraform.tfvars` or command-line versus accepting the defaults from `variables.tf` or `main.tf` (plan files). Basically, if you're using a boolean variable in the plan, anything you provide that will override this is considered a string, and not the same type as the original variable.
### Terraform Version

0.6.15
### Affected Resource(s)

variables
### Terraform Configuration Files

``` hcl
variable "boolean_a" { default = true }
variable "boolean_b" { default = false }

resource "null_resource" "boolean" {
  provisioner "local-exec" {
    command = <<-EOC
      echo "boolean_a is ${var.boolean_a}"
      echo "boolean_b is ${var.boolean_b}"
      if [ ${var.boolean_a} -eq 0 ]; then echo "boolean_a is true"; else echo "boolean_a is false"; fi
      if [ ${var.boolean_b} -eq 0 ]; then echo "boolean_b is true"; else echo "boolean_b is false"; fi
      EOC
  }

}
```

First run this plan without a `terraform.tfvars` file, then run it with a `terraform.tfvars` file specifying the exact same thing:

``` hcl
boolean_a = "true"
boolean_b = "false"
```

It should be noted that a `terraform.tfvars` file such as the following is invalid:

``` hcl
boolean_a = true
boolean_b = true
```

Results in:

``` text
invalid value "terraform.tfvars" for flag -var-file-default: Error decoding Terraform vars file: At 1:13: root.boolean_a: unknown type for string *ast.LiteralType
```

The following error ocurs when using -var 'boolean_a=true' ; requires "true" again here - resulting in a string:

``` text
The vars file should be in the format of `key = "value"`.
Decoding errors are usually caused by an invalid format.
```
### Debug Output

Without terraform.tfvars: https://gist.github.com/mengesb/0a879e53581e8c0eb97e144dd482a83f
With terraform.tfvars: https://gist.github.com/mengesb/564592c33ffa1690f5479a10ea6bdd6c
### Panic Output

No panic
### Expected Behavior

Values in `terraform.tfvars` should be treated like their boolean counterparts and behave in the same manner as when provided on command-line or in `variables.tf`/`main.tf`
### Actual Behavior

Boolean correctly intrepreted when using defaults from `main.tf` or `variables.tf`, and intrepreted as String type when intrepreting from `terraform.tfvars`
### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. Use the above plan
2. Run `terraform apply` and get a positive result. True = 1, False = 0
3. Create `terraform.tfvars` with `boolean_a = "true"` and `boolean_b = "false"
4. Run`terraform apply`and observe the`null_resource.boolean` fail as the interpreted values are now strings
### Important Factoids

True should equal True for all cases, and not intrepret it as a string. Types should be enforced.
### References
- N/A

**Comment by mengesb** [comment]
It should be noted... that this does affect all data types. Any override coming from `terraform.tfvars` or command-line are interpreted as strings. This affects booleans and numerics (float/integer)

**Comment by apparentlymart** [comment]
@mengesb thanks for this report!

As of Terraform's current release, all variables and outputs are always strings.

There is actually work underway right now to support different data types for 0.7, after which this might look better... so I expect we'll probably just let this ticket sit for a moment and then see how the situation has changed once the new type system changes are merged. The issue you describe will hopefully get fixed as part of that change, but even if it doesn't I expect this will be easier to fix once that work is complete and Terraform's internals have better support for non-string types.

**Comment by pixie79** [comment]
Hi,

I still see this same issue under 0.7 for both bool and dict data types. Is there another method to override defaults ?

Regards

Mark

**Comment by mengesb** [comment]
So I have an update to the issue per Terraform v0.7.0

The below:
variables.tf:
```hcl
variable "a" {
  type = "string"
  default = false
}
```
terraform.tfvars:
```hcl
a = true
```
produces an error `* variable a should be type string, got boolean`

There's no type boolean or bool, and produces this error: `Variable 'a' must be of type string or map - 'bool' is not a valid type`

When it is part of a MAP however, it seems that I can pass booleans and use booleans in overrides. Seems to me the fix here would be to allow for a type 'boolean' and include it in the 'first-class' data objects design

**Comment by tmatilai** [comment]
Would be great to specify a variable type to also `"boolean"` (or `"number"`). Especially as `foo = true` and `foo = false` are casted to `"1"` and `"0"` (instead of `"true/false"`) in interpolation (e.g. map access), so it depends if the variable is passed with or without the quotes.

**Comment by llarsson** [comment]
This just bit me, too.

The worst thing for me is that I am trying to pass (via a chain of modules) a value to a resource that is *really* supposed to have a boolean value in it (`multi_az` in `aws_db_instance` in my case). Having gone from a hardcoded `false` in my `main.tf` file to being told by Terraform to make it into a string (`"false"`) when I tried to turn it into a variable via my `terraform.tfvars` file just feels wrong. It's a boolean! The [source code](https://github.com/hashicorp/terraform/blob/ad57b445e91e2ee4141b284477c1b9eb4b11b820/builtin/providers/aws/resource_aws_db_instance.go#L153) thinks so, as does [the AWS API that is being used](https://github.com/hashicorp/terraform/blob/ad57b445e91e2ee4141b284477c1b9eb4b11b820/builtin/providers/aws/resource_aws_db_instance.go#L409) during invocation. Passing in a string that says `"false"` feels utterly *strange*.

When can we expect to be able to specify other types than just string, list, or map (the latter of which requires a default value to actually make Terraform believe that I was serious when I specified `map` as type)?
